### ISSUE TITLE
Attributes API with virtual attributes don't work in form builders without a value?

### ISSUE THREAD

**Issue by cannikin** [rails/rails#25787]
Playing with virtual attributes via the Attributes API in Rails 5. They don't seem to work in Form Builders unless you initialize the attribute with a value before the form builder gets ahold of it:

### Steps to reproduce

```ruby
# app/models/user.rb
class User < ApplicationRecord
  # first_name, last_name in database
  attribute :title, :string # not in DB
end

# app/controllers/users_controller.rb
class UsersController < ApplicationController
  def edit
    @user = User.find(params[:id])
  end
end
```

```erb
<%# app/views/users/edit.html.erb %>
<%= form_for @user do |f| %>
  <%= f.text_field :first_name %>
  <%= f.text_field :last_name %>
  <%= f.text_field :title %>
<% end %>
```

### Expected behavior
Should show a form with three input boxes, one each for `first_name`, `last_name` and `title`.

### Actual behavior
`ActiveModel::MissingAttributeError: missing attribute: title`

I'm able to get the form to render properly if, in the controller, I set the title to an empty string:

```ruby
class UsersController < ApplicationController
  def edit
    @user = User.find(params[:id])
    @user.title = ''
  end
end
```

I thought maybe I could do this automatically with `attribute :title, :string, :default = ''` but I get the same error. :( Maybe this is expected (and desired?) behavior? I just haven't seen it documented anywhere.

Last few items in the stack trace:

```
activerecord (5.0.0) lib/active_record/attribute_methods/read.rb:66:in `block in _read_attribute'
activerecord (5.0.0) lib/active_record/attribute_set.rb:42:in `block in fetch_value'
activerecord (5.0.0) lib/active_record/attribute.rb:196:in `value'
activerecord (5.0.0) lib/active_record/attribute_set.rb:42:in `fetch_value'
activerecord (5.0.0) lib/active_record/attribute_methods/read.rb:66:in `_read_attribute'
activerecord (5.0.0) lib/active_record/attribute_methods/read.rb:36:in `__temp__d6563737167656'
ransack (1.7.0) lib/ransack/helpers/form_builder.rb:10:in `value'
```

### System configuration
**Rails version**: 5.0.0

**Ruby version**: 2.3.1p112

**Comment by maclover7** [comment]
cc @sgrif

**Comment by kesha-antonov** [comment]
IMO it should be set automatically.
Same problem.

```
after_commit :update_rank,
    if: lambda{|x| x.should_update_rank == true}
```
I've got
```
ActiveModel::MissingAttributeError: missing attribute: should_update_rank
from /Users/kesha/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/attribute_methods/read.rb:66:in `block in _read_attribute'
```

**Comment by doits** [comment]
looks like I stumbled on this, too: #25809 (there is an executable test case)

**Comment by sled** [comment]
The reason this happens is because the non-database backed attribute is in an uninitialized state.
Here's the logic / trace behind it:

When you retrieve a record from the database, the raw result hash is passed to the [`instantiate`](https://github.com/rails/rails/blob/2797de25141098c501dd32d72af8050997cfe036/activerecord/lib/active_record/persistence.rb#L56) method of `ActiveRecord::Persistence`.

This method does not instantiate the model in a normal way e.g using `MyModel.new`, rather it allocates on a lower level and bypasses the constructor, writing directly to the instance variables.

The important thing here is that the `attributes` parameter which gets passed to `instantiate` only contains keys and values for the columns in the database. So when the attribute set is built on line [68](https://github.com/rails/rails/blob/2797de25141098c501dd32d72af8050997cfe036/activerecord/lib/active_record/persistence.rb#L68), the [attribute set](https://github.com/rails/rails/blob/2797de25141098c501dd32d72af8050997cfe036/activerecord/lib/active_record/attribute_set/builder.rb#L13) only has values defined for the database columns.

This in turn leads to an uninitialized attribute to be returned [here](https://github.com/rails/rails/blob/2797de25141098c501dd32d72af8050997cfe036/activerecord/lib/active_record/attribute_set/builder.rb#L104) here because it has no value for this attribute but the type of it is defined.

E.g `value_present == false` and `types.key?(name)` is true.

TL;DR you need to set the attribute to any value before you access it through the getter methods. 

Fun fact is though:

```
class User < ActiveRecord::Base
   attribute :title, :string
end
```

Scenario 1:
```
user = User.first
user.title => ActiveModel::MissingAttributeError: missing attribute
user.read_attribute(:title) => nil
user.title => ActiveModel::MissingAttributeError: missing attribute
user.write_attribute(:title, 'a title') 
user.title => 'a title'
```

Scenario 2:
```
user = User.first
user.assign_attributes(title: 'a title')
user.title => 'a title'
```

The workaround I use is to assign a value to custom defined attributes in `after_find`:

```
class User < ActiveRecord::Base
  attribute :title, :string

  after_find do
    # initialize to nil
    values = Hash[custom_attribute_names.map { |attr_name| [attr_name, nil] }]
    assign_attributes(values)
    # clear dirty flag
    clear_attribute_changes(custom_attribute_names)
  end

  private

  def custom_attribute_names
    self.class.attribute_names - self.class.column_names
  end
end
```

I admit, it's not the most elegant solution but it works fine in my case.

**Comment by cannikin** [comment]
@sled Dude that was an awesome breakdown, thanks!

But now the question is: is that the intended behavior (ie. you *should* always have to initialize those attributes manually)? It feels to me like if the attributes API is a valid way to define ALL the attributes of a model if you so desired (and skip the default database introspection) then the model shouldn't care if the attribute is backed by the database or not—at the very least Form Builders should display all attributes regardless. And if you set a default value on the attribute then a record should be initialized with the default (either with `Model.new` or `Model.find`). Otherwise it should be initialized as `nil`.

What happens when you try and *persist* a record where half the attributes are in the DB and half are virtual is a different story. Although it appears that you can save a record with virtual attributes just fine: virtual attributes aren't saved to the DB so it's up to you to persist them in some other way if you want, otherwise their value is lost when the object is deallocated.

**Comment by sled** [comment]
The behaviour is the same as using `.select` e.g assume a `Person` model with two DB columns: `firstname` and `lastname` then you do something like:

`person = Person.select(:firstname).first`

Calling `person.lastname` results in the same `MissingAttributeError`. 

With regular DB attributes this behaviour makes sense. If we only select a subset of the columns we don't have any data for the other columns, one might think "ok, let's set the others to `nil` then", this can bite you because you might have some code that does special things if a certain attribute is `nil`, so without knowing the "real" value in the database any code execution should be prevented that relies on the value of this attribute. The same goes for saving the subset back, you only want to save the subset of columns back to the DB without overriding the other columns.

Now what about the "Attributes API" of Rails 5? The "attributes" (i.e AttributeSet, ...) are still tied to ActiveRecord while the types are now part of ActiveModel. 

My use case for the Attributes API is a STI model which has a JSON column to store form fields, these form fields then get mapped to attributes of the ActiveRecord model, it's basically a typed version of `store_accessor`.
